### PR TITLE
RF:  ironing repo classes Part II

### PR DESCRIPTION
**Pull request by bpoldrack**

### CODE REVIEW

File: datalad/support/gitrepo.py [hunk 1]
```diff
@@ -690,6 +710,99 @@ def get_file_content(self, file_, branch='HEAD'):
...
+        if env is None:
+            env = {}
+        if git_options is None:
+            git_options = {}
+        if cmd_options is None:
+            cmd_options = {}
+        cmd_options.update({'with_exceptions': with_exceptions,
+                            'with_extended_output': True})
+
+        with self.repo.git.custom_environment(**env):
```
Comment (yarikoptic) [inline datalad/support/gitrepo.py]: This all relates to https://github.com/datalad/datalad/pull/515 and #506 I guess, or here git calls are done from within current directory?

Comment (bpoldrack) [inline datalad/support/gitrepo.py]: Nope, they are done from topdir, but can easily be done from CWD as well, as there is a dedicated option for that (like `with_exceptions` and `with_extended_output`)
I guess, I should keep an eye on  #516, right?

File: datalad/distribution/dataset.py [hunk 2]
```diff
@@ -223,6 +223,27 @@ def get_subdatasets(self, pattern=None, fulfilled=None, absolute=False,
 #        """
 #        raise NotImplementedError("TODO")
 
+    # TODO maybe needs to get its own interface
+    def remember_state(self, message, auto_add_changes=True, version=None):
```
Comment (yarikoptic) [inline datalad/distribution/dataset.py]: so it is the "save" functionality?   it is not tested BTW -- so I guess not used... worth having?

File: datalad/support/gitrepo.py [hunk 3]
```diff
@@ -64,10 +64,58 @@
...
+
+    Returns
+    -------
+    list
+    """
+    # TODO: borrow_docs!
+
+    split = kwargs.pop('split_single_char_options', True)
+    return gitpy.Git().transform_kwargs(split_single_char_options=split,
+                                        **kwargs)
```
Comment (yarikoptic) [inline datalad/support/gitrepo.py]: so what is the point of explicitly popping split_single_char_options?  or is it False by default in `transform_kwargs`?

File: datalad/support/annexrepo.py [hunk 4]
```diff
@@ -497,22 +470,14 @@ def proxy(self, git_cmd, **kwargs):
...
-        cmd_str = "git annex proxy -- %s" % git_cmd
-        # TODO: By now git_cmd is expected to be string.
-        # Figure out how to deal with a list here. Especially where and how to
-        # treat paths.
-
         if not self.is_direct_mode():
-            lgr.warning("proxy() called in indirect mode: %s" % cmd_str)
-            raise CommandNotAvailableError(cmd=cmd_str,
+            lgr.warning("proxy() called in indirect mode: %s" % git_cmd)
+            raise CommandNotAvailableError(cmd="git annex proxy",
```
Comment (yarikoptic) [inline datalad/support/annexrepo.py]: so easy to make a unittest for this one and beef up our coverage by 2 lines! ;)

File: datalad/support/gitrepo.py [hunk 5]
```diff
@@ -1116,6 +1279,17 @@ def add_submodule(self, path, name=None, url=None, branch=None):
...
         self._git_custom_command('', cmd)
 
+    def deinit_submodule(self, path, **kwargs):
+        """
+        Parameters
+        ----------
+        path
+        """
+
+        kwargs['insert_kwargs_after'] = 'deinit'
```
Comment (yarikoptic) [inline datalad/support/gitrepo.py]: to be safe, make a copy first: `kwargs = kwargs.copy()`, or use `kwargs = datalad.utils.updated(kwargs, {'insert_kwargs_after': 'deinit'})`

File: datalad/support/tests/test_annexrepo.py [hunk 6]
```diff
@@ -928,6 +917,39 @@ def test_annex_ssh(repo_path, remote_1_path, remote_2_path):
     ok_(exists(socket_2))
 
 
+@with_testrepos('basic_annex', flavors=['clone'])
+@with_tempfile(mkdir=True)
+def test_annex_remove(path1, path2):
+    ar1 = AnnexRepo(path1, create=False)
```
Comment (yarikoptic) [inline datalad/support/tests/test_annexrepo.py]: coverage reports that this test is not 'covered' -- does `@with_testrepos` doesn't return any here????

File: datalad/support/tests/test_gitrepo.py [hunk 7]
```diff
@@ -719,6 +728,68 @@ def test_get_added_files_commit_msg():
...
+        assert_in("nothing to commit, working directory clean", out)
+        eq_("", err)
+        eq_("", cm.out)
+
+    # use git_options:
+    # Note: 'path2' doesn't contain a git repository
+    with assert_raises(GitCommandError) as cm:
+        repo._gitpy_custom_call('status', git_options={'C': path2})
+        assert_in("git -C %s status" % path2, str(cm.exception))
+        assert_in("fatal: Not a git repository", str(cm.exception))
```
Comment (yarikoptic) [inline datalad/support/tests/test_gitrepo.py]: the two lines aren't executed since I guess status call right away throws GitCommandError.  So I guess then need to be dedented outside of the context management  But first store cm.exception since I Guess it would not be avaailable outside

File: datalad/distribution/dataset.py (hunk 2 above)
Comment (bpoldrack) [inline datalad/distribution/dataset.py]: Actually not part of this PR. Replaced by `save` as you mentioned. (remember: toplevel functions are methods of `Dataset`). So, it's a matter of merge. Should disappear completely.

File: datalad/support/tests/test_annexrepo.py (hunk 6 above)
Comment (bpoldrack) [inline datalad/support/tests/test_annexrepo.py]: Don't get it by now. Locally, it's working for me. Meaning: The test is actually running.

**Comment by bpoldrack** [comment]
There's something wrong with the setup of nd70. A lot of

    git-annex: external special remote protocol error, unexpectedly received "" (unable to parse   command)

I usually get this kind of error, if the datalad executables (including the special remote things) are either not in `PATH` or not build.
@yarikoptic , any chance you have a look at it?

File: datalad/support/tests/test_annexrepo.py [hunk 8]
```diff
@@ -124,6 +124,16 @@ def test_AnnexRepo_annex_proxy(src, annex_path):
     ar.set_direct_mode(True)
     ok_clean_git_annex_proxy(path=annex_path)
 
+    # annex proxy raises in indirect mode:
+    try:
+        ar.set_direct_mode(False)
+        assert_raises(CommandNotAvailableError, ar.proxy, ['git', 'status'])
+    except CommandNotAvailableError:
+        # we can't switch to indirect
```
Comment (yarikoptic) [inline datalad/support/tests/test_annexrepo.py]: how do you know?  place assert into `else:`?

**Comment by yarikoptic** [comment]
nd70 got upgraded python-git and it is all green now . so would be nice I guess to figure out why that test is not excercised.

**Comment by bpoldrack** [comment]
Apparently it actually is executed by Travis (as well as nd14 for example). See Travis' log (the errors are okay):

    datalad.support.tests.test_annexrepo.test_annex_remove ... 
    2016-06-29 12:32:12,075 [ERROR  ] 
    stderr| error: the following file has changes staged in the index:
             |     rm-test.dat
             | (use --cached to keep the file, or -f to force removal) (cmd.py:352)
    2016-06-29 12:32:12,075 [ERROR  ] 
    Failed to run ['git', '-c', 'receive.autogc=0', '-c', 'gc.auto=0', 'annex', 'proxy', '--', 'git', 'rm',
    'rm-test.dat'] under '/tmp/datalad_temp_test_annex_removeZU7viU'. Exit code=1. out= err=error:
    the following file has changes staged in the index:
    |     rm-test.dat
    | (use --cached to keep the file, or -f to force removal)
    |  (cmd.py:296)
    ok

If only the decorators would be called and the test function itself was not actually executed, there couldn't be that output. Instead these exceptions are the ones I test for. Also the file name `rm-test.dat` can only rise from this test. So, the question is: Why don't `codecov ` and `coverage` know it was executed?

**Comment by yarikoptic** [comment]
yeah... something is strange... I guess noone of us would like to dig it deeper at this point, so I will just merge for now.
